### PR TITLE
added regexpMatchScalar

### PR DESCRIPTION
**Pull request by Lyra126**
## Summary of the changes / Why this improves CrateDB

Added RegexpMatchScalar

Submitting this pull request as a solution to the following issue: https://github.com/crate/crate/issues/15117

- Created a class, implementing it similar to the existing RegexpMatchOperator.
- Instead inherited it from Scalar instead of Operator
- Had it support the use of flags.
- Updated ScalarFunctionModule.java to register it as a function

## Checklist

 - [ ] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [✓] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [✓] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by seut** [comment]
@Lyra126 
Thank you for this contribution.
Probably I confused you by my comment https://github.com/crate/crate/issues/15117#issuecomment-1909597298 that the scalar is almost equal to the existing operator. But the scalar should return the matched values, so the return type should be a array of strings ->`DataTypes.STRING_ARRAY`. 
See also the example in the original issue request, this example must work.

File: server/src/main/java/io/crate/expression/scalar/regex/RegexpMatchScalar.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        String pattern = (String) literal.value();
+        int flags = 0;
+
+        String flagsStr = null;
+        int flagSeparatorIndex = pattern.lastIndexOf('/');
+        if (flagSeparatorIndex != -1 && flagSeparatorIndex < pattern.length() - 1) {
+            pattern = pattern.substring(0, flagSeparatorIndex);
+            flagsStr = pattern.substring(flagSeparatorIndex + 1);
+            flags = parseFlags(flagsStr);
+        }
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/regex/RegexpMatchScalar.java:101]: Any optional flag would be the 3rd argument and not part of the pattern, see https://www.postgresql.org/docs/current/functions-matching.html.
To support this, a custom `toQuery(Function function, LuceneQueryBuilder.Context context)` would be needed to take the 3rd argument into account.

**Review by seut — COMMENTED** [review]
Could you also add related tests, best by extending from `ScalarTestCase` please? thanks!

**Comment by Lyra126** [comment]
Thank you for your continuous review and help with this issue! This is my first time contributing, and I appreciate the help and clarification. Let me know if I'm missing anything else.

File: server/src/main/java/io/crate/expression/scalar/regex/RegexpMatchScalar.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            if (!isGlobal(flags)) {
+                pattern = Pattern.compile(pattern, parseFlags(flags)).pattern();
+            } else{
+                throw new IllegalArgumentException("Unsupported flag: 'g'. Global flag is not supported.");
+            }
+        }
+
+        List<String> matches = new ArrayList<>();
+        java.util.regex.Matcher matcher = java.util.regex.Pattern.compile(pattern).matcher(source);
+        while (matcher.find()) {
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/regex/RegexpMatchScalar.java:97]: According to the PG documentation, it should return `null` if no match is found.
Additionally, the results of the all groups in the current match should be returned, not the first group only but for all global matches (the global flag 'g' must no be supported by this method).

File: server/src/main/java/io/crate/expression/scalar/regex/RegexpMatchScalar.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    @Override
+    public Query toQuery(Reference ref, Literal<?> literal) throws IllegalArgumentException {
+        String pattern = (String) literal.value();
+        int flags = 0;
+
+        // Check if the pattern contains flags
+        String flagsStr = null;
+        int flagSeparatorIndex = pattern.lastIndexOf('/');
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/regex/RegexpMatchScalar.java:110]: I wonder where the flags are added to the pattern by the `/` separator? I think the existing `toQuery` implementations won't work as they only support 2 arguments, a reference and a literal, but not 3.

**Comment by seut** [comment]
@Lyra126 Any updates on this, do you need additional help?

**Comment by Lyra126** [comment]
> @Lyra126 Any updates on this, do you need additional help?

Yeah, I’ve realized that I’m a bit lost and may not have the necessary experience to resolve this. I’d like to step back so that someone more familiar with the project can take over. Thank you for checking in though.

**Comment by seut** [comment]
@Lyra126 Ok thank you for trying, I'll close this then.

**Comment by bhaskar16** [comment]
Hello @seut , can I work on this please?

**Comment by seut** [comment]
@bhaskar16 Sure, go ahead, thanks! 
Btw. we stopped assigning tickets as this assignment got stale a lot of times as people did not progress on the development while also not inform us. So just go ahead on working on an issue and create a (Draft) PR asap so we can see your progress.

**Pull request closed without merging**